packages/authx/src/model/rules/IsAccessibleByRule.ts

```diff
@@ -0,0 +1,142 @@
+import { Rule } from "./Rule";
+import { Authorization } from "../Authorization";
+import { extract } from "@authx/scopes";
+
+/**
+ * Rule that handles pagination if the user is paging backwards through the data
+ * Always fetches one extra row from the data source so that the receiver can
+ * determine if there is another page.
+ *
+ * Note that this class only determines if the authorization has the right to see the entity
+ * at all. It doesn't check that the user has the right to see any specific fields of the
+ * entity, such as secrets or details. It is assumed that the class itself handles that.
+ */
+export class IsAccessibleByRule extends Rule {
+  constructor(
+    private readonly realm: string,
+    private readonly authorization: Authorization,
+    private readonly entityType: string
+  ) {
+    super();
+  }
+
+  private prepared = false;
+  private where = "";
+  private params: { [p: string]: any } = {};
+
+  private static readonly ENTITY_MAPPING_TABLE: {
+    [key: string]: { [key: string]: string };
+  } = {
+    user: {
+      userid: "entity_id"
+    },
+    authorization: {
+      userid: "user_id",
+      authorizationid: "entity_id",
+      grantid: "grant_id",
+      clientid:
+        "(SELECT client_id FROM authx.grant_record WHERE entity_id = grant_id AND replacement_record_id IS NULL)"
+    },
+    client: {
+      clientid: "entity_id"
+    },
+    credential: {
+      credentialid: "entity_id",
+      authorityid: "authority_id",
+      userid: "user_id"
+    },
+    grant: {
+      grantid: "entity_id",
+      userid: "user_id",
+      clientid: "client_id"
+    },
+    role: {
+      roleid: "entity_id"
+    }
+  };
+
+  private ensurePrepared(): void {
+    if (!this.prepared) {
+      const template = `${this.realm}:v2.${this.entityType}.(authorityid).(authorizationid).(clientid).(credentialid).(grantid).(roleid).(userid):r....`;
+
+      const fixedIds: { [key: string]: string }[] = [];
+      let allAccess = false;
+
+      for (const scope of this.authorization.scopes) {
+        const extracted = extract(template, [scope]);
+
+        if (extracted.length == 1) {
```

Comment:
I _think_ that one way to handle these and reduce some logical indirection overall would be to use `this.entityType` to select a more limited template. For example, using the following scope template against all available scopes:

```js
scopes.extract(
  `${this.realm}:v2.user.......(user_id):r....`,
  [
    // any large number of complex scopes
  ]
)
```

will return only matches with a relevant `user_id`. Similarly, templates that contain more than one parameter will already be broken down into their normal disjunctive form (ors of ands). Just an idea!

Reply:
I think a simpler approach would be to just use all the results from the extraction? Or is there some case where that wouldn't work?

Reply:
Ah yes - this works correctly too; my follow-up comment was just a suggestion for being able to potentially remove the logic that loops through the results and makes sure each resulting scope is relevant to `this.entityType` (since it would already be filtered) and you could just focus on the SQL generation loop. But this is perfectly fine too!